### PR TITLE
Handle virtual object creation in a background job

### PR DESCRIPTION
**Pull request by mjgiarlo**
Connects to #336
Deployment blocked by sul-dlss/puppet#4470

* Add ActiveRecord and Postgres as dependencies. This makes it possible to reliably stash information about background jobs in a persistent store, which will be used to run virtual object creation in the background.
* Include `factory_bot_rails` as a dependency which will make it easier to spin
up fixture-like factories when testing with database entities later.
* Create a model and a table for background job results. The model holds any output from the job, an HTTP status code, and a processing status.
* Add factory_bot_rails to the test suite. This allows us to more efficiently test database-backed entities, specifically `BackgroundJobResults`.
* Add new controller and view for background job results.
* Use a background job to create virtual objects. Use the new job from the `CreateVirtualObjectsController`. In the job, record progress and results in the application database

## Why was this change made?

To be able to process large batches of virtual objects outside of the request/response cycle.

## Was the API documentation (openapi.json) updated?

Yes.

### CODE REVIEW

File: app/controllers/virtual_objects_controller.rb [hunk 1]
```diff
@@ -2,27 +2,21 @@
 
 # Create virtual objects
 class VirtualObjectsController < ApplicationController
-  before_action :validate_params!, only: :create
+  before_action :validate_params!, :create_background_job_result, only: :create
```
Comment (jcoyne) [inline app/controllers/virtual_objects_controller.rb]: Is there a reason we need to put this in a callback?  I think it's easier to read if it's just in the method.

File: app/controllers/virtual_objects_controller.rb [hunk 2]
```diff
@@ -2,27 +2,21 @@
...
-      errors << ConstituentService.new(parent_druid: parent_id).add(child_druids: child_ids)
-    rescue ActiveFedora::ObjectNotFoundError, Rubydora::FedoraInvalidRequest, Dor::Exception => e
-      errors << { parent_id => [e.message] }
-    end
-
-    return render_error(errors: errors, status: :unprocessable_entity) if errors.any?
-
-    head :no_content
+    CreateVirtualObjectsJob.perform_later(virtual_objects: create_params[:virtual_objects],
+                                          background_job_result_id: @result.id)
```
Comment (jcoyne) [inline app/controllers/virtual_objects_controller.rb]: I think you can pass `background_job_result: @result` and it will serialize/deserialize for you using GlobalId.

File: app/jobs/create_virtual_objects_job.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+# Create virtual objects in the background
+class CreateVirtualObjectsJob < ApplicationJob
+  queue_as :default
+
+  attr_reader :background_job_result
+
+  # @param [Array] virtual_objects an array of hashes representing a batch of virtual objects
+  # @param [Integer] background_job_result_id identifier of a background job result to store status info
+  def perform(virtual_objects:, background_job_result_id:)
+    background_job_result = BackgroundJobResult.find(background_job_result_id)
```
Comment (jcoyne) [inline app/jobs/create_virtual_objects_job.rb]: Avoid this by using GlobalId.

File: app/models/background_job_result.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+
+# Database-backed model to hold results of long-running jobs
+class BackgroundJobResult < ApplicationRecord
+  enum status: { pending: 0, processing: 1, complete: 2 }
```
Comment (jcoyne) [inline app/models/background_job_result.rb]: Enums in active record are sort of a pain in the ass.  as it's not understandable to anyone using SQL.  How about using a postgresql ENUM (or even a string column) instead?

File: spec/support/auth_helper.rb [hunk 5]
```diff
@@ -4,4 +4,14 @@ module AuthHelper
   def login
     allow(controller).to receive(:check_auth_token)
   end
+
+  def jwt
```
Comment (jcoyne) [inline spec/support/auth_helper.rb]: This really deserves to be its own PR.

Comment (mjgiarlo) [reply spec/support/auth_helper.rb]: @jcoyne 	OK: https://github.com/sul-dlss/dor-services-app/pull/424

Will remove this commit when I push changes next.

**Comment by mjgiarlo** [comment]
@jcoyne :speech_balloon: 
> The PR message says you updated openapi.json, but I don't see that in the changed files.

Included now.

File: app/controllers/virtual_objects_controller.rb (hunk 1 above)
Comment (mjgiarlo) [reply app/controllers/virtual_objects_controller.rb]: I went back and forth on that. Which I'm doing again now. :smile:

File: app/controllers/virtual_objects_controller.rb (hunk 2 above)
Comment (mjgiarlo) [reply app/controllers/virtual_objects_controller.rb]: @jcoyne 	OK

File: app/jobs/create_virtual_objects_job.rb (hunk 3 above)
Comment (mjgiarlo) [reply app/jobs/create_virtual_objects_job.rb]: K

File: app/models/background_job_result.rb (hunk 4 above)
Comment (mjgiarlo) [reply app/models/background_job_result.rb]: Changed. Thanks for the tip!

File: db/migrate/20190917215521_create_background_job_results.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+class CreateBackgroundJobResults < ActiveRecord::Migration[5.2]
+  def up
+    execute <<-SQL
+      CREATE TYPE background_job_result_status AS ENUM (
```
Comment (jcoyne) [inline db/migrate/20190917215521_create_background_job_results.rb]: Great! We should port this to workflow-service.

File: openapi.json [hunk 7]
```diff
@@ -249,6 +253,45 @@
...
+        ],
+        "summary": "View results of a background job",
+        "description": "Used to allow the application run long-running processes out of the request/response cycle",
+        "operationId": "background_job_results#show",
+        "responses": {
+          "200": {
+            "description": "OK"
+          },
+          "202": {
+            "description": "Accepted"
```
Comment (jcoyne) [inline openapi.json]: Can you say that "this is returned when the job has been queued, but has not completed processing.

File: openapi.json [hunk 8]
```diff
@@ -249,6 +253,45 @@
...
+      "get": {
+        "tags": [
+          "jobs"
+        ],
+        "summary": "View results of a background job",
+        "description": "Used to allow the application run long-running processes out of the request/response cycle",
+        "operationId": "background_job_results#show",
+        "responses": {
+          "200": {
+            "description": "OK"
```
Comment (jcoyne) [inline openapi.json]: "This is returned when the job has completed processing successfully"

File: openapi.json [hunk 9]
```diff
@@ -258,12 +301,12 @@
         "description": "Combines a parent object with child objects to create an object like an Atlas (composed of several images)",
         "operationId": "virtual_objects#create",
         "responses": {
-          "204": {
-            "description": "objects created"
-          },
-          "422": {
-            "description": "unable to process request"
+          "201": {
+            "description": "virtual object creation job created"
```
Comment (jcoyne) [inline openapi.json]: Can you describe the `Location` header here? https://swagger.io/docs/specification/describing-responses/#headers

File: openapi.json (hunk 7 above)
Comment (mjgiarlo) [reply openapi.json]: done, pushed

File: openapi.json (hunk 8 above)
Comment (mjgiarlo) [reply openapi.json]: done, pushed

File: openapi.json (hunk 9 above)
Comment (mjgiarlo) [reply openapi.json]: yep, coming soon

Comment (mjgiarlo) [reply openapi.json]: Pushed.
